test/gtest/ucs/test_type.cc

```diff
@@ -138,6 +142,145 @@ UCS_TEST_F(test_type, pack_float) {
+        for (int c = 0; c < num; c++) {
+            tt.emplace_back([&]() {
+                unsigned seed = time(0);
+                for (int i = 0; i < 1000000 / num; i++) {
+                    if ((rand_r(&seed) % 256) < writers) {
+                        w();
+                    } else {
+                        r();
+                    }
+                }
+            });
+        }
+
+
+        for (auto &t : tt) {
+            t.join();
+        }
+        auto end = std::chrono::high_resolution_clock::now();
+        std::chrono::duration<double> elapsed = end - start;
+
+        UCS_TEST_MESSAGE << elapsed.count() * 1000 << " ms " << name << " "
+                         << std::to_string(num) << " threads "
+                         << std::to_string(writers) << " writers per 256 ";
+    }
+
+    void measure(const std::function<void()> &r,
+                 const std::function<void()> &w, const std::string &name)
+    {
+        int m = std::thread::hardware_concurrency();
+        std::vector<int> threads = {1, 2, 4, m};
+        std::vector<int> writers_per_256 = {1, 25, 128, 250};
+
+        for (auto t : threads) {
+            for (auto writers : writers_per_256) {
+                measure_one(t, writers, r, w, name);
+            }
+        }
+    }
+};
+
+UCS_TEST_F(test_rwlock, lock) {
+    ucs_rwlock_t lock = UCS_RWLOCK_STATIC_INITIALIZER;
+
+    ucs_rwlock_read_lock(&lock);
+    EXPECT_EQ(-EBUSY, ucs_rwlock_write_trylock(&lock));
+
+    ucs_rwlock_read_lock(&lock); /* second read lock should pass */
+
+    int write_taken = 0;
+    std::thread w([&]() {
+        ucs_rwlock_write_lock(&lock);
+        write_taken = 1;
+        ucs_rwlock_write_unlock(&lock);
+    });
+    sleep();
+    EXPECT_FALSE(write_taken); /* write lock should wait for read lock release */
+
+    ucs_rwlock_read_unlock(&lock);
+    sleep();
+    EXPECT_FALSE(write_taken); /* first read lock still holding lock */
+
+    int read_taken = 0;
+    std::thread r1([&]() {
+        ucs_rwlock_read_lock(&lock);
+        read_taken = 1;
+        ucs_rwlock_read_unlock(&lock);
+    });
+    sleep();
+    EXPECT_FALSE(read_taken); /* read lock should wait while write lock is waiting */
+
+    ucs_rwlock_read_unlock(&lock);
+    sleep();
+    EXPECT_TRUE(write_taken); /* write lock should be taken */
+    w.join();
+
+    sleep();
+    EXPECT_TRUE(read_taken); /* read lock should be taken */
+    r1.join();
+
+    EXPECT_EQ(0, ucs_rwlock_write_trylock(&lock));
+    read_taken = 0;
+    std::thread r2([&]() {
+        ucs_rwlock_read_lock(&lock);
+        read_taken = 1;
+        ucs_rwlock_read_unlock(&lock);
+    });
+    sleep();
+    EXPECT_FALSE(read_taken); /* read lock should wait for write lock release */
+
+    ucs_rwlock_write_unlock(&lock);
+    sleep();
+    EXPECT_TRUE(read_taken); /* read lock should be taken */
+    r2.join();
+}
+
+UCS_TEST_F(test_rwlock, perf) {
+    ucs_rwlock_t lock = UCS_RWLOCK_STATIC_INITIALIZER;
+    measure(
```

Comment:
This is a good performance test, but it must change some state to guarantee the lock correctness. This is the whole purpose of litmus test. For instance, these functions may perform some simple math calculations and we check the invariant:
```C
    // Invariant: counter2 is 2 times bigger than counter1
    int counter1 = 1;
    int counter2 = 2;
    measure(
            [&]() {
                ucs_rwlock_read_lock(&lock);
                UCS_ASSERT_EQ(counter1 * 2, counter2);
                ucs_rwlock_read_unlock(&lock);
            },
            [&]() {
                ucs_rwlock_write_lock(&lock);
                counter1 += counter1;
                counter2 += counter2;
                if (counter2 > 100000) {
                    counter1 = 1;
                    counter2 = 2;
                }
                ucs_rwlock_write_unlock(&lock);
            },
```

Reply:
I'm not sure this is a good idea. This test doesn't guarantee lock correctness, it will very easily give a false negative result.

Reply:
Well, testing correctness of MT algorithms is hard topic. You're right about false negatives results, that's ok. That's actually a nature of litmus tests: when they pass, it does not guarantee that your algorithm is 100% correct, but when they fail - it's obviously broken. Personally I catch tons of MT issues with the help of litmus tests.
If you propose some other way of testing correctness - let's discuss that. What I'm proposing is well established industry practise, and I'm sure it's better to have it than no testing at all.

Moreover, the example that I provided is just scratching the surface. We should also consider adding tests with nested locks, try-locks, etc